### ISSUE TITLE
range_tombstones_compaction_test with CDC feature enabled has a performance degredation

### ISSUE THREAD

**Issue by fgelcer** [scylladb/scylla#9744]
during rolling upgrade, we are installing as the base version Scylla 4.5, and then we perform a huge set of queries/tests, and for each step (while we upgrade, node by node) we re-run all these queries/tests again.

we saw that this specific test is taking too long, so i have created a [branch](https://github.com/fgelcer/scylla-cluster-tests/tree/upgrade_test_cdc_tombstone_tests_only) that will allow us to run only this query/test, to measure how long it is taking.

while running with the last build that passed (http://downloads.scylladb.com/unstable/scylla/master/rpm/centos/2021-11-18T04:00:29Z/scylla.repo) we got the following results:
```
< t:2021-12-05 21:53:16,807 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 11 seconds
< t:2021-12-05 22:37:00,793 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 288 seconds
< t:2021-12-05 22:59:32,186 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 372 seconds
< t:2021-12-05 23:14:56,588 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 450 seconds
< t:2021-12-05 23:33:02,682 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 471 seconds
< t:2021-12-05 23:57:53,019 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 521 seconds
< t:2021-12-06 00:23:29,059 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 509 seconds
```

but when running with the last build we had running for CentOS8 upgrade test (http://downloads.scylladb.com/unstable/scylla/master/rpm/centos/latest/scylla.repo that resolves to http://downloads.scylladb.com/unstable/scylla/master/rpm/centos/2021-11-28T03:59:37Z/scylla.repo) we have the following results:
```
< t:2021-12-05 22:01:00,890 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 11 seconds
< t:2021-12-05 22:48:20,535 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 466 seconds
< t:2021-12-05 23:20:03,456 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 874 seconds
< t:2021-12-05 23:48:12,203 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 1044 seconds
< t:2021-12-06 00:22:50,584 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 1193 seconds
< t:2021-12-06 00:57:04,255 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 1101 seconds
< t:2021-12-06 01:41:11,862 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 1102 seconds
```

it represents an increase of between 60% up to 150% in the time these tests are taking to run.

results for the test with the "older" build:
[sct_runner.log](https://cloudius-jenkins-test.s3.amazonaws.com/63d61176-40b0-4354-8f2e-a9e5e2765e6b/20211206_005501/sct-runner-63d61176.tar.gz)
[db_logs](https://cloudius-jenkins-test.s3.amazonaws.com/63d61176-40b0-4354-8f2e-a9e5e2765e6b/20211206_005501/db-cluster-63d61176.tar.gz)

the "newer" build logs:
[sct_runner.log](https://cloudius-jenkins-test.s3.amazonaws.com/1248418b-9d93-4735-907a-f9dd737716e7/20211206_021523/sct-runner-1248418b.tar.gz)
[db_logs](https://cloudius-jenkins-test.s3.amazonaws.com/1248418b-9d93-4735-907a-f9dd737716e7/20211206_021523/db-cluster-1248418b.tar.gz)

**Comment by slivne** [comment]
@haaawk - this is a regression - there is a week difference of a week between the two commits and we need to check whats the source.

QA are working to extract this to a dtest so it will be possible to look at this.

Roy suspected this regression is also visible between 4.5.0 <--> 4.5.2 - need to look at the commits that went it

**Comment by fgelcer** [comment]
> @haaawk - this is a regression - there is a week difference of a week between the two commits and we need to check whats the source.
> 
> QA are working to extract this to a dtest so it will be possible to look at this.
> 
> Roy suspected this regression is also visible between 4.5.0 <--> 4.5.2 - need to look at the commits that went it

[here](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/fabio/job/rolling-upgrade-cdc-timeout/job/only_tombstone_tests_upgrade_centos8/4/) i'm running with upgrade from 4.5 to 4.5.2 (expect 4.5.2 to be the last stable one, as i'm running with the official 4.5 repo).

will start working now on porting this to dtest (will take the existing test, and will enable CDC in it)

**Comment by haaawk** [comment]
@kbr- can this be somehow related to CDC streams?
@fgelcer @roydahan is CDC actually enabled on any table or just turning the feature on causes the performance degradation?

**Comment by fgelcer** [comment]
> @kbr- can this be somehow related to CDC streams? @fgelcer @roydahan is CDC actually enabled on any table or just turning the feature on causes the performance degradation?

i'm working on comparing the same scenario in dtest (to be easier to reproduce), but so far not many results to share...

the original test (the rolling upgrade tests for 4.5, 4.6 and master) run with CDC enabled to ALL tables, so here i took the ones that have the "worst" performance, and running only 2 of them (2 tables only), with CDC enabled

**Comment by roydahan** [comment]
> @kbr- can this be somehow related to CDC streams?
> @fgelcer @roydahan is CDC actually enabled on any table or just turning the feature on causes the performance degradation?

In short to your question, the CDC is actually enabled, we configured many tables we have in the test to use it.
@fgelcer please include an example of the table.

just to make clear, when we enabled this feature as part of the test few months ago it was working fine for at least 3 months.

**Comment by fgelcer** [comment]
from the test, here are the tables, populations and queries:
```
{
            'name': "range_tombstones_test: Test deletion by 'composite prefix' (range tombstones)",
            'create_tables': ["""CREATE TABLE range_tombstones_test (
                            k int,
                            c1 int,
                            c2 int,
                            v1 int,
                            v2 int,
                            PRIMARY KEY (k, c1, c2)
                        )"""],
            'truncates': [],
            'inserts': ["INSERT INTO range_tombstones_test (k, c1, c2, v1, v2) VALUES (%d, %d, %d, %d, %d)" % (
                i, j, k, (i * 4) + (j * 2) + k, (i * 4) + (j * 2) + k) for i in range(0, 5) for j in range(0, 2)
                for k
                in range(0, 2)],
            'queries': ["SELECT v1, v2 FROM range_tombstones_test where k = %d" % 0,
                        "SELECT v1, v2 FROM range_tombstones_test where k = %d" % 1,
                        "SELECT v1, v2 FROM range_tombstones_test where k = %d" % 2,
                        "SELECT v1, v2 FROM range_tombstones_test where k = %d" % 3,
                        "SELECT v1, v2 FROM range_tombstones_test where k = %d" % 4,
                        "DELETE FROM range_tombstones_test WHERE k = %d AND c1 = 0" % 0,
                        "DELETE FROM range_tombstones_test WHERE k = %d AND c1 = 0" % 1,
                        "DELETE FROM range_tombstones_test WHERE k = %d AND c1 = 0" % 2,
                        "DELETE FROM range_tombstones_test WHERE k = %d AND c1 = 0" % 3,
                        "DELETE FROM range_tombstones_test WHERE k = %d AND c1 = 0" % 4,
                        "SELECT v1, v2 FROM range_tombstones_test WHERE k = %d" % 0,
                        "SELECT v1, v2 FROM range_tombstones_test WHERE k = %d" % 1,
                        "SELECT v1, v2 FROM range_tombstones_test WHERE k = %d" % 2,
                        "SELECT v1, v2 FROM range_tombstones_test WHERE k = %d" % 3,
                        "SELECT v1, v2 FROM range_tombstones_test WHERE k = %d" % 4,
                        "#REMOTER_SUDO nodetool flush",
                        "SELECT v1, v2 FROM range_tombstones_test WHERE k = %d" % 0,
                        "SELECT v1, v2 FROM range_tombstones_test WHERE k = %d" % 1,
                        "SELECT v1, v2 FROM range_tombstones_test WHERE k = %d" % 2,
                        "SELECT v1, v2 FROM range_tombstones_test WHERE k = %d" % 3,
                        "SELECT v1, v2 FROM range_tombstones_test WHERE k = %d" % 4],
            'results': [[[x, x] for x in range(0 * 4, (0 + 1) * 4)],
                        [[x, x] for x in range(1 * 4, (1 + 1) * 4)],
                        [[x, x] for x in range(2 * 4, (2 + 1) * 4)],
                        [[x, x] for x in range(3 * 4, (3 + 1) * 4)],
                        [[x, x] for x in range(4 * 4, (4 + 1) * 4)],
                        [],
                        [],
                        [],
                        [],
                        [],
                        [[x, x] for x in range(0 * 4 + 2, (0 + 1) * 4)],
                        [[x, x] for x in range(1 * 4 + 2, (1 + 1) * 4)],
                        [[x, x] for x in range(2 * 4 + 2, (2 + 1) * 4)],
                        [[x, x] for x in range(3 * 4 + 2, (3 + 1) * 4)],
                        [[x, x] for x in range(4 * 4 + 2, (4 + 1) * 4)],
                        None,
                        [[x, x] for x in range(0 * 4 + 2, (0 + 1) * 4)],
                        [[x, x] for x in range(1 * 4 + 2, (1 + 1) * 4)],
                        [[x, x] for x in range(2 * 4 + 2, (2 + 1) * 4)],
                        [[x, x] for x in range(3 * 4 + 2, (3 + 1) * 4)],
                        [[x, x] for x in range(4 * 4 + 2, (4 + 1) * 4)]],
            'min_version': '',
            'max_version': '',
            'skip': ''},
{
            'name': "range_tombstones_compaction_test: Test deletion by 'composite prefix' "
                    "(range tombstones) with compaction",
            'create_tables': ["""CREATE TABLE range_tombstones_compaction_test (
                            k int,
                            c1 int,
                            c2 int,
                            v1 text,
                            PRIMARY KEY (k, c1, c2)
                        )"""],
            'truncates': [],
            'inserts': ["INSERT INTO range_tombstones_compaction_test (k, c1, c2, v1) VALUES (0, %d, %d, '%s')" % (
                c1, c2, '%i%i' % (c1, c2)) for c1 in range(0, 4) for c2 in range(0, 2)],
            'queries': ["#REMOTER_SUDO nodetool flush",
                        "DELETE FROM range_tombstones_compaction_test WHERE k = 0 AND c1 = 1",
                        "#REMOTER_SUDO nodetool flush",
                        "#REMOTER_SUDO nodetool compact",
                        "SELECT v1 FROM range_tombstones_compaction_test WHERE k = 0"],
            'results': [None,
                        [],
                        None,
                        None,
                        [['%i%i' % (c1, c2)] for c1 in range(0, 4) for c2 in range(0, 2) if c1 != 1]],
            'min_version': '',
            'max_version': '',
            'skip': ''}
```

**Comment by kbr-** [comment]
> previous version (a.k.a. base_version) for upgrade is scylla-4.5

Which scylla-4.5 exactly was used in each run?

> i have reverted (in a test for 4.5.2 release, where base_version is 4.4) the CDC enabling commit to this test (running the very same test with CDC disabled) and then the test passed (while with CDC enabled it doesn't).

Do I understand correctly, the compared paths were:
1. upgrade from 4.4 to 4.5.2 with CDC disabled
2. upgrade from 4.5 to master with CDC enabled with master from 18.11
3. upgrade from 4.5 to master with CDC enabled with master from 28.11

2 and 3 are described in the original post, and 1 you checked just now?

I don't think it makes sense to compare 4.4->4.5.2 with 4.5->master if we want to determine whether CDC is to blame, too many variables are changing.

To understand whether it's really CDC that causes the regression we need results from the following:
4. upgrade from 4.5 to master with CDC disabled with master from 18.11
5. upgrade from 4.5 to master with CDC disabled with master from 28.11

So the only difference between 2, 3 vs 4, 5 is that CDC is enabled/disabled.

We know there was regression between 2. and 3., but there may also be a regression between 4. and 5., if so, CDC is probably not to blame.

**Comment by fgelcer** [comment]
the whole investigation started because of 1, but we can put this aside for now (although it seems like the version is affected by the same issue), so we need to focus on 2 and 3... i will try to get a version working to run 4 and 5.

i agree that there might be other things in, and as the problematic "test" inside rolling upgrade test is related to tombstones (when CDC is enabled), then it could be a combination of factors...

**Comment by fgelcer** [comment]
@kbr- , here what you asked...
upgrading from 4.5 to master (the exact same builds as in the issue description) have almost same duration for these queries:

build from `2021-11-18` - [logs here](https://cloudius-jenkins-test.s3.amazonaws.com/4b269235-4f57-4549-9869-ea1f84095c8d/20211207_230234/db-cluster-4b269235.tar.gz) or your step 4:
```
12 seconds
196 seconds
195 seconds
251 seconds
250 seconds
263 seconds
248 seconds
```

build from `2021-11-28` [logs here](https://cloudius-jenkins-test.s3.amazonaws.com/87cda035-f8aa-4dae-be46-988bbf796124/20211207_225852/db-cluster-87cda035.tar.gz), or your step 5:
```
11 seconds
192 seconds
190 seconds
239 seconds
241 seconds
249 seconds
253 seconds
```

about the dtest, so far results are not very conclusive (couldn't reproduce it in dtest)
i will try to isolate it even more in the upgrade tests and share here once and if i have more details.

@kbr- , LMK what else you need here to investigate this issue.

**Comment by kbr-** [comment]
Thanks @fgelcer. So it does seem there is no regression with CDC disabled, but there is a regression with CDC enabled.

On the other hand we only have a single test run for each scenario. Can we be sure that the slow run (described in the first post) wasn't a random one-time non-reproducible event? Or were there multiple such runs and all of them were slow?

**Comment by fgelcer** [comment]
@kbr- it is consistently reproducible in master, and i have tried (for 4.5.2 release) to run the very same rolling upgrade, but disabling the CDC feature, and it passed...
you can see in master that all upgrade tests are failing (the majority with timeout, cause by this increased time to run these queries).

**Comment by kbr-** [comment]
Ok. Probably the next best step would be to bisect using the daily builds from 18.11 to 28.11, for each build run the upgrade test from 4.5 to this build with CDC enabled and see where the regression starts earliest, so we can pinpoint the exact commit (unless you manage to find a faster reproducer that doesn't require upgrade - that would be great). I briefly looked at the git log and didn't see any CDC related changes that could cause something like that, hence I'm suspecting it's something else (maybe again a TWCS compaction related thingy?)

I'm working on another task right now, but maybe next week I'll be able to focus on this, unless @haaawk finds someone who's currently free and could investigate further.

If you want to work on this further @fgelcer then one idea that I would check would be to try the test with CDC disabled but change the table from the test to use TimeWindowCompactionStrategy (and the 28.11 build).

**Comment by fgelcer** [comment]
@kbr- , maybe something that could help...
i'm seeing this error message, after each run of the `range_tombstone_test` runs:
`Search for idx_token error`

```
< t:2021-12-08 09:48:37,312 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 12 seconds
< t:2021-12-08 09:53:23,734 f:upgrade_test.py l:534  c:UpgradeTest          p:INFO  > pre-test - Run stress workload before upgrade
< t:2021-12-08 10:03:02,018 f:upgrade_test.py l:571  c:UpgradeTest          p:INFO  > Step1 - Upgrade First Node 
< t:2021-12-08 10:20:38,640 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 190 seconds
< t:2021-12-08 10:20:39,131 f:upgrade_test.py l:471  c:UpgradeTest          p:DEBUG > Search for idx_token error. Step Step1 - Upgrade First Node  - after upgraded one node
< t:2021-12-08 10:20:39,332 f:upgrade_test.py l:601  c:UpgradeTest          p:INFO  > Step2 - Upgrade Second Node 
< t:2021-12-08 10:40:39,939 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 301 seconds
< t:2021-12-08 10:40:40,545 f:upgrade_test.py l:471  c:UpgradeTest          p:DEBUG > Search for idx_token error. Step Step2 - Upgrade Second Node  - after upgraded two nodes
< t:2021-12-08 10:40:40,665 f:upgrade_test.py l:622  c:UpgradeTest          p:INFO  > Step3 - Rollback Second Node 
< t:2021-12-08 10:46:41,276 f:upgrade_test.py l:630  c:UpgradeTest          p:INFO  > Step4 - Verify data during mixed cluster mode 
< t:2021-12-08 10:53:07,618 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 353 seconds
< t:2021-12-08 10:55:47,464 f:upgrade_test.py l:471  c:UpgradeTest          p:DEBUG > Search for idx_token error. Step Step4 - Verify data during mixed cluster mode 
< t:2021-12-08 10:55:47,761 f:upgrade_test.py l:640  c:UpgradeTest          p:INFO  > Step5 - Upgrade rest of the Nodes 
< t:2021-12-08 11:10:15,798 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 452 seconds
< t:2021-12-08 11:10:16,766 f:upgrade_test.py l:471  c:UpgradeTest          p:DEBUG > Search for idx_token error. Step Step5 - Upgrade rest of the Nodes 
< t:2021-12-08 11:38:27,846 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 653 seconds
< t:2021-12-08 11:38:28,498 f:upgrade_test.py l:471  c:UpgradeTest          p:DEBUG > Search for idx_token error. Step Step5 - Upgrade rest of the Nodes 
< t:2021-12-08 12:05:55,203 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 756 seconds
< t:2021-12-08 12:05:55,920 f:upgrade_test.py l:471  c:UpgradeTest          p:DEBUG > Search for idx_token error. Step Step5 - Upgrade rest of the Nodes 
< t:2021-12-08 12:05:56,329 f:upgrade_test.py l:651  c:UpgradeTest          p:INFO  > Step6 - Verify stress results after upgrade 
< t:2021-12-08 12:05:56,332 f:upgrade_test.py l:658  c:UpgradeTest          p:INFO  > Step7 - Upgrade sstables to latest supported version 
< t:2021-12-08 12:09:27,764 f:upgrade_test.py l:678  c:UpgradeTest          p:INFO  > Step8 - Run stress and verify after upgrading entire cluster 
< t:2021-12-08 12:09:36,205 f:upgrade_test.py l:723  c:UpgradeTest          p:INFO  > Step9 - Search for errors that we filter during the test 
< t:2021-12-08 12:09:36,587 f:upgrade_test.py l:471  c:UpgradeTest          p:DEBUG > Search for idx_token error. Step Step9 - Search for errors that we filter during the test 
< t:2021-12-08 12:09:37,268 f:upgrade_test.py l:471  c:UpgradeTest          p:DEBUG > Search for idx_token error. Step Step9 - Search for errors that we filter during the test 
< t:2021-12-08 12:09:37,909 f:upgrade_test.py l:471  c:UpgradeTest          p:DEBUG > Search for idx_token error. Step Step9 - Search for errors that we filter during the test 
< t:2021-12-08 12:09:39,313 f:upgrade_test.py l:471  c:UpgradeTest          p:DEBUG > Search for idx_token error. Step Step9 - Search for errors that we filter during the test 
< t:2021-12-08 12:09:42,412 f:upgrade_test.py l:747  c:UpgradeTest          p:INFO  > Step10 - Verify that gemini and cdc stressor are not failed during upgrade
```

full logs can be found [here](https://cloudius-jenkins-test.s3.amazonaws.com/a0b25d04-99ec-40f0-87b3-e49bd98dd964/20211208_121943/db-cluster-a0b25d04.tar.gz)

**Comment by fgelcer** [comment]
just a second before we start the upgrade of the 1st node, i see this in the logs:
```
< t:2021-12-08 10:03:02,018 f:events_device.py l:172  c:sdcm.sct_events.events_device p:DEBUG > <EventsDevice name='EventsDevice-1' pid=133 parent=52 unknown daemon>: add filter (DbEventsFilter type=RUNTIME_ERROR line=Failed to load schema) with uuid=2058f6ec-0046-4e55-bdb0
-b40db6c5bb7b
< t:2021-12-08 10:03:02,019 f:events_device.py l:172  c:sdcm.sct_events.events_device p:DEBUG > <EventsDevice name='EventsDevice-1' pid=133 parent=52 started daemon>: add filter (DbEventsFilter type=RUNTIME_ERROR line=Failed to load schema) with uuid=2058f6ec-0046-4e55-bdb0
-b40db6c5bb7b
< t:2021-12-08 10:03:02,019 f:events_device.py l:172  c:sdcm.sct_events.events_device p:DEBUG > <EventsDevice name='EventsDevice-1' pid=133 parent=52 started daemon>: add filter (DbEventsFilter type=RUNTIME_ERROR line=Failed to load schema) with uuid=2058f6ec-0046-4e55-bdb0
-b40db6c5bb7b
< t:2021-12-08 10:03:02,018 f:upgrade_test.py l:571  c:UpgradeTest          p:INFO  > Step1 - Upgrade First Node 
< t:2021-12-08 10:03:02,020 f:events_device.py l:172  c:sdcm.sct_events.events_device p:DEBUG > <EventsDevice name='EventsDevice-1' pid=133 parent=52 started daemon>: add filter (DbEventsFilter type=DATABASE_ERROR line=cql_server - exception while processing connection: seastar::nested_exception (seastar::nested_exception)) with uuid=1d4c2121-7b06-4c2a-bfb4-b3e6dd9d6e93
< t:2021-12-08 10:03:02,020 f:events_device.py l:172  c:sdcm.sct_events.events_device p:DEBUG > <EventsDevice name='EventsDevice-1' pid=133 parent=52 unknown daemon>: add filter (DbEventsFilter type=DATABASE_ERROR line=cql_server - exception while processing connection: seastar::nested_exception (seastar::nested_exception)) with uuid=1d4c2121-7b06-4c2a-bfb4-b3e6dd9d6e93
< t:2021-12-08 10:03:02,020 f:events_device.py l:172  c:sdcm.sct_events.events_device p:DEBUG > <EventsDevice name='EventsDevice-1' pid=133 parent=52 started daemon>: add filter (DbEventsFilter type=DATABASE_ERROR line=cql_server - exception while processing connection: seastar::nested_exception (seastar::nested_exception)) with uuid=1d4c2121-7b06-4c2a-bfb4-b3e6dd9d6e93
```

**Comment by aleksbykov** [comment]
Additional information, which could help.

During job: https://jenkins.scylladb.com/job/scylla-master/job/rolling-upgrade/job/rolling-upgrade-ami-test/72/execution/node/67/log/?consoleFull, After first node was upgraded from 4.5.2 -> 4.7.dev: 0.20211212.b0a8c153f
,  suspicious query trigger next error:
```
12:26:55  < t:2021-12-10 05:26:52,896 f:fill_db_data.py l:3191 c:UpgradeTest          p:INFO  > Update CDC tables for test "range_tombstones_test: Test deletion by 'composite prefix' (range tombstones)" in 0 seconds
12:30:13  < t:2021-12-10 05:30:12,478 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 2021-12-10 05:30:12.476: (PrometheusAlertManagerEvent Severity.NORMAL) period_type=begin event_id=99ad9171-d55e-42b7-8440-f014a922d651: alert_name=HighLatencies node=10.0.2.42 start=2021-12-10T05:30:04.791Z end=2021-12-10T05:34:04.791Z description=10.0.2.42 has average high latency for more than 5 minutes. updated=2021-12-10T05:30:11.448Z state=active fingerprint=e3f81b11344861aa labels={'alertname': 'HighLatencies', 'by': 'instance', 'instance': '10.0.2.42', 'monitor': 'scylla-monitor', 'scheduling_group_name': 'statement', 'severity': '1'}
12:30:13  < t:2021-12-10 05:30:12,488 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 2021-12-10 05:30:12.479: (PrometheusAlertManagerEvent Severity.NORMAL) period_type=begin event_id=65491168-46d7-4ec1-a725-cb5e8a663fdb: alert_name=HighLatencies node=10.0.0.118 start=2021-12-10T05:30:04.791Z end=2021-12-10T05:34:04.791Z description=10.0.0.118 has average high latency for more than 5 minutes. updated=2021-12-10T05:30:11.448Z state=active fingerprint=fb60c8769c4a17fa labels={'alertname': 'HighLatencies', 'by': 'instance', 'instance': '10.0.0.118', 'monitor': 'scylla-monitor', 'scheduling_group_name': 'statement', 'severity': '1'}
12:30:13  < t:2021-12-10 05:30:12,517 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 2021-12-10 05:30:12.483: (PrometheusAlertManagerEvent Severity.NORMAL) period_type=begin event_id=d0eaeac1-d498-4a85-9fbc-a683631c7298: alert_name=HighLatencies node=10.0.3.58 start=2021-12-10T05:30:04.791Z end=2021-12-10T05:34:04.791Z description=10.0.3.58 has average high latency for more than 5 minutes. updated=2021-12-10T05:30:11.448Z state=active fingerprint=ff9d781190c56d00 labels={'alertname': 'HighLatencies', 'by': 'instance', 'instance': '10.0.3.58', 'monitor': 'scylla-
monitor', 'scheduling_group_name': 'statement', 'severity': '1'}

12:41:24  < t:2021-12-10 05:41:19,497 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 021-12-10T05:41:19+00:00 rolling-upgrade-master-centos-db-node-cf07edf8-1 ! WARNING |  [shard 0] service_level_controller - update_from_distributed_data: failed to update configuration for more than  90 seconds : exceptions::read_timeout_exception (Operation timed out for system_distributed.service_levels - received only 0 responses from 1 CL=ONE.)
12:41:24  < t:2021-12-10 05:41:19,500 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 2021-12-10 05:41:19.493 <2021-12-10 05:41:19.000>: (DatabaseLogEvent Severity.ERROR) period_type=one-time event_id=6e10617e-cbf5-44fc-979c-7ce9b6d31200: type=DATABASE_ERROR regex=Exception  line_number=185881 node=Node rolling-upgrade-master-centos-db-node-cf07edf8-1 [52.18.14.209 | 10.0.0.84] (seed: False)
```
Same repeated after second node was upgraded.
During mixed mode, when cluster have old and new version c-s command failed with next errors:
```
15:17:02  < t:2021-12-10 08:17:00,010 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > java.io.IOException: Operation x10 on key(s) [4d323035504c36373030]: Error executing: (OperationTimedOutException): [10.0.3.58/10.0.3.58:9042] Timed out waiting for server response
15:17:02  < t:2021-12-10 08:17:00,170 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 2021-12-10 08:17:00.166: (CassandraStressLogEvent Severity.CRITICAL) period_type=one-time event_id=506cbcac-4b42-4ba6-a28d-64c7fbe3ddf4: type=OperationOnKey regex=Operation x10 on key\(s\) \[ line_number=241109 node=Node rolling-upgrade-master-centos-loader-node-cf07edf8-1 [52.17.64.10 | 10.0.0.157] (seed: False)
15:17:02  < t:2021-12-10 08:17:00,170 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > java.io.IOException: Operation x10 on key(s) [32333439343034333131]: Error executing: (OperationTimedOutException): [10.0.3.58/10.0.3.58:9042] Timed out waiting for server response
15:17:02  < t:2021-12-10 08:17:00,176 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > 2021-12-10 08:17:00.168: (CassandraStressLogEvent Severity.CRITICAL) period_type=one-time event_id=506cbcac-4b42-4ba6-a28d-64c7fbe3ddf4: type=OperationOnKey regex=Operation x10 on key\(s\) \[ line_number=241117 node=Node rolling-upgrade-master-centos-loader-node-cf07edf8-1 [52.17.64.10 | 10.0.0.157] (seed: False)
15:17:02  < t:2021-12-10 08:17:00,176 f:file_logger.py  l:85   c:sdcm.sct_events.file_logger p:INFO  > java.io.IOException: Operation x10 on key(s) [3732503932354d323930]: Error executing: (OperationTimedOutException): [10.0.3.58/10.0.3.58:9042] Timed out waiting for server response
```

```

**Comment by slivne** [comment]
I know we are running bisect - but its not clear what is the current range under suspicion - can you please clear this up

**Comment by kbr-** [comment]
Current status according to my best knowledge is:
- we managed to do some bisecting with @fgelcer; the regression probably happened between daily builds 19.11 and 21.11, with commits 4b1bb26d5 and f3d5b2b2b respectively (4b1bb26d5 is probably good, f3d5b2b2b is bad). The commits in between are:
```
f3d5b2b2b Merge "Add effective_replication_map factory" from Benny
c94e6f856 compaction: Merge GC writer into regular compaction writer
f8c91bdd1 Update debugging.md
22fa7ecf9 Update compaction_controller.md
eed3e9570 effective_replication_map: clear_gently when destroyed
cd0061dcb database: shutdown keyspaces
1e259665f test: cql_test_env: stop view_update_generator before database shuts down
866e1b847 effective_replication_map_factory: try cloning replication map from shard 0
0a3d66839 tools: schema_loader: start a sharded erm_factory
23e1344b7 storage_service: use erm_factory to create effective_replication_map
cb240ffba keyspace: use erm_factory to create effective_replication_map
6754e6ca2 effective_replication_map: erase from factory when destroyed
8a6fbe800 effective_replication_map_factory: add create_effective_replication_map
ecba37dbf effective_replication_map: enable_lw_shared_from_this
f4f41e290 effective_replication_map: define factory_key
5947de767 keyspace: get a reference to the erm_factory
1d7556d09 main: pass erm_factory to storage_service
242043368 main: pass erm_factory to storage_proxy
3fed73e7c locator: add effective_replication_map_factory
3c0fec6b1 storage_proxy: paxos_response_handler::prune: demote write timeout error printout to debug level
5f7ee2e13 test: sstable_compaction_test: fix twcs_reshape_with_disjoint_set_test by using a non-coarse timestamp resolution
1dd08e367 test, cross-shard-barrier: Increase stall detector period
87a212fa5 memtable-sstable: Fix indentation in table::try_flush_memtable_to_sstable.
6df07f7ff memtable-sstable: Convert table::try_flush_memtable_to_sstable to coroutines.
0f404c727 test: raft: randomized_nemesis_test: better RPC message receiving implementation
a51529dd1 protocol_servers: strengthen guarantees of listen_addresses()
```
- there is a hypothesis that the regression in the test run time comes from `nodetool compact`, which takes a lot of time to compact tables written by cassandra-stress in the upgrade test. @jul-stas is benchmarking compaction and I heard he's onto something. We may see some interesting results soon.

**Comment by jul-stas** [comment]
I tried to reproduce it locally, on custom built:
- 44f4ea38c5b70e80dfeb2962d21447b3c53dda49 (23.11.2021) and
- 4b1bb26d5a064b55dae8a906a7e45f7322b9ff44 (17.11.2021).

That is: create the same schema as in SCT **but without CDC** -> load it with 1E6 rows of similar data -> flush -> delete 10% of data with a prefix key -> flush -> compact. The results are unclear to me.
- With STCS on the table: no difference was observed and results were consistent.
- With TWCS on the table: there is a performance difference in the compaction step (up to 2x!), but the variance of measurements is high.

Note that I did it on a dev build, on my desktop, and measured wall time. It might mean something, may not.
I'll re-run the tests with bigger dataset, bigger timestamp span, real CDC, and a more real-life cluster, but if I had to guess - something related to TWCS would be my no. 1 trail at the moment.

**Comment by slivne** [comment]
I don't understand why we are comparing master if we know the regression is on 4.5 as well - that would make the bisect x10 faster

**Comment by slivne** [comment]
@roydahan / @fgelcer / @jul-stas - I don't understand whats going on this is a mess

The regression was first reported in upgrade of 4.5.0..4.5.2 - if this is true the bisect is totally off

If this is not true and the regression is only on master - please clarify it -

**Comment by kbr-** [comment]
The original post in this issue reported a regression in upgrade test from 4.5 to master. I'm not aware of any regression between 4.5.0 and 4.5.2.

**Comment by fgelcer** [comment]
on the last job of 4.5 it got even worst...
```
< t:2021-12-22 22:51:24,544 f:fill_db_data.py l:3156 c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 17 seconds
< t:2021-12-23 02:10:40,210 f:fill_db_data.py l:3156 c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 9599 seconds
```

it ran only 2 times, and then the test timed out and stopped.
the 2nd query happened during:
```
Step1 - Upgrade First Node
```

**Comment by ilya-rarov** [comment]
Faced the same issue during the rolling upgrade job for CentOs 8 ([this build](https://jenkins.scylladb.com/view/master/job/scylla-master/job/rolling-upgrade/job/rolling-upgrade-centos8-test/106/)). The job reached the timeout and was eventually aborted too .
The upgrade was running from version 4.6.rc1-0.20211208.542394c82
```
[2021-12-17T05:27:07.337Z] < t:2021-12-17 05:26:59,732 f:fill_db_data.py l:3191 c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 106 seconds
[2021-12-17T06:23:23.022Z] < t:2021-12-17 06:23:20,139 f:fill_db_data.py l:3191 c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 856 seconds
[2021-12-17T07:12:19.803Z] < t:2021-12-17 07:12:14,494 f:fill_db_data.py l:3191 c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 1484 seconds
[2021-12-17T08:04:13.759Z] < t:2021-12-17 08:04:11,470 f:fill_db_data.py l:3191 c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 1653 seconds
[2021-12-17T08:51:36.221Z] < t:2021-12-17 08:51:31,225 f:fill_db_data.py l:3191 c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 1444 seconds
[2021-12-17T09:29:00.122Z] < t:2021-12-17 09:28:44,184 f:fill_db_data.py l:3191 c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 1458 seconds
```
[DB cluster logs](https://cloudius-jenkins-test.s3.amazonaws.com/0be1edd5-eba8-4eea-9131-dff4c25c5729/20211217_100510/db-cluster-0be1edd5.tar.gz)
Unfortunately, the log collector process was aborted too so there is no ZIP with SCT logs.

**Comment by fgelcer** [comment]
during the last upgrade AMI for 4.5 we have even a worst regression:
```
Test deletion by 'composite prefix' (range tombstones) with compaction" in 17 seconds
Test deletion by 'composite prefix' (range tombstones) with compaction" in 2948 seconds
Test deletion by 'composite prefix' (range tombstones) with compaction" in 3584 seconds
```

[db logs](https://cloudius-jenkins-test.s3.amazonaws.com/40f76763-5cd0-4569-ba01-111d87020980/20211224_022747/db-cluster-40f76763.tar.gz)
[sct logs](https://cloudius-jenkins-test.s3.amazonaws.com/40f76763-5cd0-4569-ba01-111d87020980/20211224_022747/sct-runner-40f76763.tar.gz)

**Comment by fgelcer** [comment]
i'm trying right now to run the same test (only tombstone tests, with CDC enabled), but without doing the upgrades (all the loads, tests and queries), starting from 4.5.2...
let's see if this would be a simpler reproducer

**Comment by slivne** [comment]
Based on @fgelcer there is a regression between 4.5.0 --> 4.5.2 (now on 4.5.3 build) - we need to bisect this and find the issue

**Comment by slivne** [comment]
@roydahan / @fgelcer we need info on what job needs to be run to reproduce the issue (reliably enough if possible 1 out of 3 runs is good 1 out of 1 is better) - lets start from jenkins job and afterwords with whoever works with you provide additional info

**Comment by kbr-** [comment]
@fgelcer can you point me to a Jenkins job which shows the aforementioned regression between 4.5.0 and 4.5.2?
The only job I found that you used was the one which checked 4.5.2->master upgrade path, which we used to do the bisecting.

**Comment by slivne** [comment]
Lets the split the issue into two - there is a the regression in 4.5.0 <-> 4.5.2 and the regression that julius found in https://github.com/scylladb/scylla/issues/9744#issuecomment-993887719

**Comment by avikivity** [comment]
It's possible that we're creating lots of effective_replication_map:s without noticing, and that this slows down compaction.

**Comment by fgelcer** [comment]
> @fgelcer can you point me to a Jenkins job which shows the aforementioned regression between 4.5.0 and 4.5.2? The only job I found that you used was the one which checked 4.5.2->master upgrade path, which we used to do the bisecting.

[here is the job](https://jenkins.scylladb.com/job/scylla-4.5/job/rolling-upgrade/job/rolling-upgrade-ami-test/93/)

**Comment by fgelcer** [comment]
i have tried running the limited test (tombstone tests only, with CDC enabled) with a version is supposed to have the slow results [here](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/fabio/job/rolling-upgrade-cdc-timeout/job/only_tombstone_tests_upgrade_centos8_previous_passed/) (without doing upgrade)...
the result turns out to be "good", and leads me to thing that, it could be something related to the "mixed" cluster while upgrading... maybe related to tombstones that changed between these versions, or even something else (like hinted handoff) slowing the queries...

**Comment by kbr-** [comment]
> [here is the job](https://jenkins.scylladb.com/job/scylla-4.5/job/rolling-upgrade/job/rolling-upgrade-ami-test/93/)

This test shows a single upgrade from 4.4.8 to 4.5.3. To show a regression between 4.5.0 and 4.5.2 (or 4.5.3) we would need something to compare with, e.g. 4.4.8->4.5.0 upgrade. I see no such run however. Even better, the last two successful runs on this job don't even have `Ran queries for test "range_tombstones_compaction_test"` in the logs - so it seems that the test was introduced recently. The testing framework itself is changing with the Scylla versions being tested. You can't deduce anything from that.

**Comment by kbr-** [comment]
> i have tried running the limited test (tombstone tests only, with CDC enabled) with a version is supposed to have the slow results here (without doing upgrade)...
the result turns out to be "good", and leads me to thing that, it could be something related to the "mixed" cluster while upgrading... maybe related to tombstones that changed between these versions, or even something else (like hinted handoff) slowing the queries...

By removing the upgrade step, you made the test much shorter. Just look at the timestamps between different `Ran queries for test "range_tombstones_compaction_test"` log messages.
For example:
```
[2022-01-02T11:37:25.800Z] < t:2022-01-02 11:37:25,292 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 10 seconds
...
[2022-01-02T11:54:12.199Z] < t:2022-01-02 11:53:58,866 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 109 seconds
```
there is a 26 minute difference between these log lines.
Then:
```
[2022-01-02T11:56:21.900Z] < t:2022-01-02 11:56:18,571 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 120 seconds
```
only 3 minutes later.

The longer the test runs, the more data gets inserted, and the more time it takes to run `nodetool compact`, which is the slowest operation in the test.


By making the test much shorter, you caused cassandra-stress to insert much less data, which caused the compactions to run much faster obviously.

You're comparing runs where data was inserted for hours with runs where it was inserted for minutes. **You can't deduce anything from that.**

**Comment by fgelcer** [comment]
> Ran queries for test "range_tombstones_compaction_test"

@kbr- 
this nice print was only added recently, but the CDC enabled has already passed (from 4.4.x to 4.5.0)

**Comment by fgelcer** [comment]
> > i have tried running the limited test (tombstone tests only, with CDC enabled) with a version is supposed to have the slow results here (without doing upgrade)...
> > the result turns out to be "good", and leads me to thing that, it could be something related to the "mixed" cluster while upgrading... maybe related to tombstones that changed between these versions, or even something else (like hinted handoff) slowing the queries...
> 
> By removing the upgrade step, you made the test much shorter. Just look at the timestamps between different `Ran queries for test "range_tombstones_compaction_test"` log messages. For example:
> 
> ```
> [2022-01-02T11:37:25.800Z] < t:2022-01-02 11:37:25,292 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 10 seconds
> ...
> [2022-01-02T11:54:12.199Z] < t:2022-01-02 11:53:58,866 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 109 seconds
> ```
> 
> there is a 26 minute difference between these log lines. Then:
> 
> ```
> [2022-01-02T11:56:21.900Z] < t:2022-01-02 11:56:18,571 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 120 seconds
> ```
> 
> only 3 minutes later.
> 
> The longer the test runs, the more data gets inserted, and the more time it takes to run `nodetool compact`, which is the slowest operation in the test.
> 
> By making the test much shorter, you caused cassandra-stress to insert much less data, which caused the compactions to run much faster obviously.
> 
> You're comparing runs where data was inserted for hours with runs where it was inserted for minutes. **You can't deduce anything from that.**

so in that case, i could add a sleep ~25 min instead of the upgrade?
i'm now trying to test same scenario, but without the c-s running (i remember doing that already, but i didn't find the results of it)

**Comment by kbr-** [comment]
>  this nice print was only added recently, but the CDC enabled has already passed (from 4.4.x to 4.5.0)

I'm not convinced other things didn't change that could slow down the test itself, since we already know that the test has changed somehow (the prints).

By the way, the last `Ran queries for test...` logs happen after `Teardown is starting`... the fact that we're shutting down everything as we're trying to obtain some sensible results doesn't help.

We would have to use the same version of SCT and compare 4.4.8->4.5.0 VS 4.4.8->4.5.3.

If we ever want to get any sensible results, we need a stable version of SCT that does not crash for random reasons in 9 out of 10 runs, but actually tests the thing we want to test. Then use this single stable version to compare different Scylla versions.

**Comment by fgelcer** [comment]
> test has changed somehow (the prints).

yes, it was failing, so we added these prints to be easier to debug the failure...

**Comment by fgelcer** [comment]
> If we ever want to get any sensible results, we need a stable version of SCT that does not crash for random reasons in 9 out of 10 runs, but actually tests the thing we want to test. Then use this single stable version to compare different Scylla versions.

so in this case, how do you suggest we create a reproducer for this? i'm trying, but having too many failures from both Scylla and SCT... and as the man in the middle, i'm not having any success to have a good and simple reproducer...

**Comment by kbr-** [comment]
> so in that case, i could add a sleep ~25 min instead of the upgrade?

If you are sure that upgrade takes ~25 minutes.

But to be honest, this is a very inefficient way to debug these problems anyway. We're wasting so much time. Trying to use SCT is a pain in the neck, when I tried to bisect it kept failing because it couldn't setup the nodes, and other reasons which I don't understand. **Why don't we have a stable version of SCT**? Imagine that our customers had to use nightly builds of Scylla and upgrade every day. This is basically what we do here when trying to test Scylla with SCT.

Rant aside... We already know the problematic step: compaction. We know that compaction slows down. There is a simple test we can devise from that:
1. insert a lot of data to CDC enabled tables.
2. start measuring time.
2. run `nodetool compact`.
3. stop measuring time.

Run this test on 4.5.0, 4.5.3, master. Compare.
No upgrades needed.

@jul-stas started doing it but is right now on vacation. When he comes back I hope he continues.

>  i'm now trying to test same scenario, but without the c-s running (i remember doing that already, but i didn't find the results of it)

That's a completely different scenario. The c-s running is what makes the test slow, because it inserts data that is then being compacted. Compaction is the longest step in the test. Not the selects.
In other words, you'll just see each "Ran queries..." finish very quickly.

**Comment by kbr-** [comment]
> That's a completely different scenario. The c-s running is what makes the test slow, because it inserts data that is then being compacted. Compaction is the longest step in the test. Not the selects.

Let me make it more clear.

The tables you create and select explicitly in the test **don't matter** for the test speed.

The only thing that matters here and slows the test is the data inserted by cassandra-stress to tables created by cassandra-stress (which you don't select in the test). Then `nodetool compact` compacts every keyspace, including these c-s tables, and this is what takes the longest.

**Comment by kbr-** [comment]
> insert a lot of data to CDC enabled tables.

And I mean **a lot**. In the test, we have cassandra-stress inserting data for **hours**.

**Comment by fgelcer** [comment]
> It's possible that we're creating lots of effective_replication_map:s without noticing, and that this slows down compaction.

@kbr- , Avi here suggested that we may be creating too many maps... WDYT? have you checked for it? is there a way to check the amount of maps we are creating? how can we measure the impact of it?

**Comment by kbr-** [comment]
> @kbr- , Avi here suggested that we may be creating too many maps... WDYT? have you checked for it? is there a way to check the amount of maps we are creating? how can we measure the impact of it?

This is a plausible explanation for the 4.5.3 -> master regression.
But this doesn't explain 4.5.0 -> 4.5.3 regression (if there is any, which I'm not convinced still).

But I don't know much about those `effective_replication_map`s and how to measure them. @bhalevy might know more.

Still, I think we should focus on getting a useful reproducer first (that shows the regression, preferably is fast, and works every time), only then looking for the cause and fixing the bug.

**Comment by fgelcer** [comment]
> Still, I think we should focus on getting a useful reproducer first (that shows the regression, preferably is fast, and works every time), only then looking for the cause and fixing the bug.

so help us out with any ideas on how to do that...
i have made few tries, and none of them are "fast"... i wanted to be sure we are not depending on the upgrade (that makes everything slower)... but still, my current reproducer takes several hours...

**Comment by kbr-** [comment]
> > Still, I think we should focus on getting a useful reproducer first (that shows the regression, preferably is fast, and works every time), only then looking for the cause and fixing the bug.
> 
> so help us out with any ideas on how to do that... i have made few tries, and none of them are "fast"... i wanted to be sure we are not depending on the upgrade (that makes everything slower)... but still, my current reproducer takes several hours...

First, upgrade tests wouldn't be so bad if we were sure that starting such a test will give us useful results after say 3 hours. In the meantime I can do something else.
What happens now is that after 10 minutes I get an email that the test failed. Which distracts me from my current work - I need to check what happened (SCT failed to setup a cluster) and restart the test. Do this 10 times a day and maybe some run actually manages to pass the 3 hours mark, so I get one set of useful results.
For this we need a stable and tested version of SCT.

Second, we can use the test I described here: https://github.com/scylladb/scylla/issues/9744#issuecomment-1004048480. @jul-stas did this already (very similar): https://github.com/scylladb/scylla/issues/9744#issuecomment-993887719 and obtained useful results. And he only inserted 1M rows, which isn't that much, so it shouldn't take that long! (if you insert them concurrently)

**Comment by fgelcer** [comment]
> 1. insert a lot of data to CDC enabled tables.
> 2. start measuring time.
> 3. run `nodetool compact`.
> 4. stop measuring time.
> 
> Run this test on 4.5.0, 4.5.3, master. Compare. No upgrades needed.

i'm trying to reproduce these steps in the following jobs:
[on 4.5.0](https://jenkins.scylladb.com/job/scylla-4.5/job/rolling-upgrade/job/reproducer_to_kamil/2/)
[on 4.5.3](https://jenkins.scylladb.com/job/scylla-4.5/job/rolling-upgrade/job/reproducer_to_kamil_4.5.3/2/)

@kbr- FYI

**Comment by fgelcer** [comment]
on 4.5.0 the compaction took ~11 minutes:
```
(InfoEvent Severity.NORMAL): message=Measuring time before compaction = 1641499282.8179984
(InfoEvent Severity.NORMAL): message=Measuring time after compaction = 1641499984.021778
```

**Comment by roydahan** [comment]
@kbr- and @fgelcer both of you do things I don’t understand.

@kbr- SCT changes if you use the tip of the master like any other repo, but why don’t you work with a branch/tag.
Of course things may not go smoothly when working with master on releases that were branched more than a year ago.

@fgelcer from the point of reproducing the cig consistently that we were long ago, we don’t need new ideas, we need to start narrow down factors and deduce the problem to the minimal reproducer we can find. I already explained you that removing the upgrade part made the test much shorter, so to check if it’s related to the amount of data that is being written during the test, what is the result of the same thing without stress and without upgrade? Does it reproduce on 4.5.3?
If not, with upgrade and without stress, does it reproduce?
(IIUC you already did the test with stress and without upgrade on 4.5.3 and it didn’t reproduce?).
If you have the answer to these 3 simple tests, we have an answer if there is a simplest repoducer.

**Comment by kbr-** [comment]
> @kbr- SCT changes if you use the tip of the master like any other repo, but why don’t you work with a branch/tag.
Of course things may not go smoothly when working with master on releases that were branched more than a year ago.

We did use a single branch for bisecting with Fabio.
The test would still often fail on the setup stage.
So I guess we were just unlucky and took a bad commit to create the branch.

So which commit of SCT should I use for bisecting Scylla 4.5 -> Scylla master upgrades? Is there a stable and reliable version of SCT which manages to successfully setup a cluster and run a test at least, say, 9 out of 10 times?

A nice thing for SCT would be some release engineering.
For example, suppose we release SCT 1.0 stable version. It's designed to support Scylla 4.4, 4.5, and master. It may also support 4.3 and earlier but there is no guarantee.
If a bug is found, it is backported to the 1.x branch and a minor version is eventually released, SCT 1.1.
New features are developed on SCT master branch (e.g. new cloud provider support) and we reduce the risk of introducing bugs to the 1.x branch.
We can extend the 1.x branch with new tests if we want, but we don't touch the core framework; tests themselves should be independent and if one test is buggy it should not affect another test and definitely not the core.
Perhaps tests should even be in another repo, and mostly independent from "SCT core framework" version, so they can potentially work with different SCT versions.
When we want to start testing Scylla in new environments such as with new cloud providers, or using new features of the framework (say different / improved nemeses), we branch SCT 2.x.
And so on.

I know that SCT does branch together with Scylla versions, but I don't know how that works. Are these branched versions stable? And this still doesn't give us a dedicated solution for testing Scylla master. It would also be nice to have a stable version for testing master branch. It should work most of the time - I think Scylla doesn't change the method used to deploy it very often, any new config parameters always have defaults, and so on, so even if Scylla master changes, this stable version of SCT should not require an update (not a big one at least).
Maybe I should use SCT for Scylla 4.5 if I want a stable version to test master? But even if it's reasonably stable there's still a problem - this "SCT 4.5" will stop being maintained and bugs will stop being fixed as soon as Scylla 4.6 is released (or am I wrong?)

**Comment by roydahan** [comment]
@kbr- This is not the right place for this discussion, but it's much simpler than you think.

To bisect scylla 4.5 to master we can use stable master point or even branch-4.6.
I don't know what the failures you had, so please try to share in SCT issue/s.

**Comment by jul-stas** [comment]
Gentlemen, can you give me two SHAs (as narrow and certain as possible ATM) to continue my experiment? A good one and a bad one, ofc.
We're chasing something tricky and volatile, so let's keep it in some order, please.

**Comment by kbr-** [comment]
> Gentlemen, can you give me two SHAs (as narrow and certain as possible ATM) to continue my experiment?

There are two suspected regressions, so two SHAs are not enough. Although I'm still not a believer in 4.5.0->4.5.3 regression :P

I'm gonna give you 4 SHAs:
4.5.0: 47be33a104b59bf69704b0a67d2e5f7eb0427d1d
4.5.3: 406b4bce8da76910e115706906ed7f3332b6f32d
Daily build 19.11: 4b1bb26d5a064b55dae8a906a7e45f7322b9ff44
Daily build 21.11: f3d5b2b2b07c036d3074cf1e646da65ae5e84cd1

We already know that there's probably a regression between 19.11 and 21.11. These commits are not far apart from each other so it should be possible to bisect to an exact commit and learn exactly what the regression is.

I suspect there is no significant regression between 4.5.3 and 19.11 daily build in compaction times. Please confirm.

There is a suspected regression between 4.5.0 and 4.5.3. Please check.

**Comment by fgelcer** [comment]
@kbr- , after my previous test where i skipped the upgrade, @roydahan suggested to add more time of load, as the issue didn't reproduce.
using Roy's suggestion, and instead of upgrading, my test now does `sleep(1200)` (20 minutes) the issue does reproduce:

```
# https://s3.amazonaws.com/downloads.scylladb.com/unstable/scylla/branch-4.5/rpm/centos/2021-10-31T08:03:13Z/scylla.repo

< t:2022-01-12 16:44:00,563 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 12 seconds
< t:2022-01-12 17:22:53,638 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 267 seconds
< t:2022-01-12 17:51:33,164 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 470 seconds
< t:2022-01-12 18:01:03,751 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 529 seconds
< t:2022-01-12 18:36:52,165 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 779 seconds
< t:2022-01-12 19:12:53,395 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 912 seconds
< t:2022-01-12 19:47:44,419 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 880 seconds



# http://downloads.scylladb.com/unstable/scylla/master/rpm/centos/2021-11-19T04:00:55Z/scylla.repo

< t:2022-01-12 16:41:11,482 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 10 seconds
< t:2022-01-12 17:23:48,588 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 542 seconds
< t:2022-01-12 18:07:13,131 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 1355 seconds
< t:2022-01-12 18:38:50,350 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 1857 seconds
< t:2022-01-12 19:38:02,891 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 2053 seconds
< t:2022-01-12 20:30:35,230 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 1944 seconds
< t:2022-01-12 21:23:11,909 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 1948 seconds
```

sadly the newer build (4.5.3) logs weren't collected, but in the job you can see these results above
[4.5.3 job](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/fabio/job/rolling-upgrade-cdc-timeout/job/only_tombstone_tests_upgrade_centos8_previous_passed/43/)
[4.5.0 job](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/fabio/job/rolling-upgrade-cdc-timeout/job/only_tombstone_tests_upgrade_centos8/11/)

the test is in [this branch on my fork](https://github.com/fgelcer/scylla-cluster-tests/tree/upgrade_test_cdc_tombstone_tests_only)

it still takes some time (could be reduced in the amount of steps in the main test, or smaller sleep time), but the scenario is simpler than before, as it doesn't depend on upgrading.

**Comment by fgelcer** [comment]
re-run the same job with upgrade to 4.5.1 and the result is very similar to the 4.5.0:
```
# 4.5.1

23:28:47  < t:2022-01-17 21:28:47,421 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 11 seconds
00:07:47  < t:2022-01-17 22:07:41,618 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 281 seconds
00:36:52  < t:2022-01-17 22:36:49,857 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 504 seconds
00:47:12  < t:2022-01-17 22:47:02,624 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 575 seconds
01:23:57  < t:2022-01-17 23:23:52,918 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 851 seconds
02:00:39  < t:2022-01-18 00:00:30,497 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 982 seconds
02:36:48  < t:2022-01-18 00:36:42,857 f:fill_db_data.py l:312  c:UpgradeTest          p:INFO  > Ran queries for test "range_tombstones_compaction_test: Test deletion by 'composite prefix' (range tombstones) with compaction" in 963 seconds
```

[4.5.1 job](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/fabio/job/rolling-upgrade-cdc-timeout/job/only_tombstone_tests_upgrade_centos8/12)

**Comment by kbr-** [comment]
> 4.5.3 job
4.5.0 job

I followed the two links and obtained the db-cluster logs for nodes for both links - 4.5.3 as well:
```
[2022-01-12T21:46:02.771Z] Uploaded to https://cloudius-jenkins-test.s3.amazonaws.com/b35bdb3c-86c7-402f-a6aa-32f6fa27bef4/20220112_214541/db-cluster-b35bdb3c.tar.gz
```



For the "4.5.0 job" logs, I checked the logs of node 0-1 and looked for "starting ..." and found:
```
2022-01-12T16:10:14+00:00 rolling-upgrade-upgrade--centos-8-db-node-0d1a6ec6-0-1 !    INFO | Scylla version 4.5.3-0.20211223.c8f14886d with build-id 9a5b504c51cbe8feb1217517d6977f7793b2971e starting ...
```

Did the same for "4.5.3 job" and found:
```
2022-01-12T16:10:02+00:00 rolling-upgrade-upgrade--centos-8-db-node-b35bdb3c-0-1 !    INFO | Scylla version 4.5.3-0.20211223.c8f14886d with build-id 9a5b504c51cbe8feb1217517d6977f7793b2971e starting ...
```

so apparently the two jobs used the same Scylla version and even the same package?

**Comment by roydahan** [comment]
Let's take out all 4.5.X related comments from this issue.
@fgelcer created a document for 4.5 bisect and it claims to show that the regression is between 4.5.0 to 4.5.1 (IIRC).
We can continue there and open a new issue regarding 4.5.

**Comment by haaawk** [comment]
It is very suspicious @roydahan. There are only 6 commits between 4.5.1 and 4.5.2:
![image](https://user-images.githubusercontent.com/1311149/150337458-d102fb43-f8c9-4707-b0c4-f573a2ad62db.png)

https://github.com/scylladb/scylla/compare/scylla-4.5.1...scylla-4.5.2

None of them looks like it could affect the compaction.

**Comment by roydahan** [comment]
There was a progress on investigating the issue on 4.5 but it may be applicable also to master.
For now it looks like the results aren't related to a specific scylla version and it could be a problem with GCP or something environmental we need to understand.

**Comment by haaawk** [comment]
@jul-stas I think that means you can park this task for now and move to other items in your queue.

**Comment by roydahan** [comment]
Yes, it's totally not clear what's going on there.
I'm looking into metrics, but for now it's assigned to @fgelcer.

**Comment by roydahan** [comment]
@fgelcer need to find the issue.

**Comment by kbr-** [comment]
The decision was to remove the test?
It seems there is a problem and we hide it?

**Comment by fgelcer** [comment]
> The decision was to remove the test? It seems there is a problem and we hide it?

we removed one of the stress tools, and it was mentioned that, since it was added, there was performance degradation... @roydahan maybe can share more details about it, but the tests are still there

**Comment by roydahan** [comment]
This issue became cumbersome and has invalid parts.
We found out that the cdcstresser tool introduce a flakiness that is not related to upgrade at all.
We are creating a separate task to check the long time it takes to query the range tombstones while the stressed is running.

**Comment by kbr-** [comment]
Ok, that makes sense. Thanks.
